### PR TITLE
refs #360: Fix dialog_caption size

### PR DESCRIPTION
**Pull request by balloonpopper**
Fixes https://github.com/carenalgas/popochiu/issues/360

### CODE REVIEW

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_caption/dialog_caption.gd [hunk 1]
```diff
@@ -2,16 +2,41 @@ extends PopochiuDialogText
...
+		for current_character in range(0, rich_text_label.text.length()):
+
+			var ThisChar = rich_text_label.text[current_character]
+			var ThisLine = rich_text_label.get_character_line(current_character)
+			if rich_text_label.get_character_line(current_character) > current_line_number:
+				current_line_number += 1
+				if rich_text_label.text[current_character-1] == " ":
+					rich_text_label.text[current_character-1] = "\n"
+				elif rich_text_label.text[current_character-1] != "\n":
+					rich_text_label.text = rich_text_label.text.left(current_character) + "\n" + rich_text_label.text.right(-current_character)
```
Comment (mapedorr) [inline addons/popochiu/engine/objects/gui/components/dialog_text/dialog_caption/dialog_caption.gd:34]: This line is too long. Could you split it so it doesn't go past column 100?

**Review by mapedorr — CHANGES_REQUESTED** [review]
I'm asking for a couple of changes. I still have to devote some time to test the behavior of the component.

**Comment by balloonpopper** [comment]
The changes to the _calculate_size function were because the existing code would split the text from this command onto 2 lines for a 100px text box, when the text was only about 60px wide when tested with (from memory) dialog_overhead.
```
await C.get_OfficerBlonde().say("It's free.")
```

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_caption/dialog_caption.gd (hunk 1 above)
Comment (balloonpopper) [reply addons/popochiu/engine/objects/gui/components/dialog_text/dialog_caption/dialog_caption.gd]: Fixed

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_text.gd [hunk 2]
```diff
@@ -79,7 +79,6 @@ func play_text(props: Dictionary) -> void:
 	await _modify_size(msg, props.position)
 	
 	# Assign the text and align mode
-	msg = "[color=%s]%s[/color]" % [props.color.to_html(), msg]
```
[the remark this answers is not in the document]
Comment (balloonpopper) [reply addons/popochiu/engine/objects/gui/components/dialog_text/dialog_text.gd:82]: Duplicated to the overhead and portrait functions.

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_caption/dialog_caption.gd [hunk 3]
```diff
@@ -4,14 +4,40 @@ extends PopochiuDialogText
 #region Private ####################################################################################
 func _modify_size(msg: String, _target_position: Vector2) -> void:
 	var _size := await _calculate_size(msg)
-	
+
 	# Define size and position (before calculating overflow)
+	rich_text_label.size.x = _size.x
 	rich_text_label.size.y = _size.y
+	rich_text_label.position.x = (get_viewport_rect().size.x/2) - (_size.x /2)
```
Comment (mapedorr) [inline addons/popochiu/engine/objects/gui/components/dialog_text/dialog_caption/dialog_caption.gd:11]: I think the ideal approach in this case is not to modify the width (`size.x`) and X position of the **RichTextLabel**. Considering how this **Control** is configured in Godot, what we should do is ensure that the value defined in the `wrap_width` property is assigned to the `size.x` property of the **RichTextLabel** in the `_ready()` function. To control the Y position of the component, Godot's anchors can be used.

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_caption/dialog_caption.gd [hunk 4]
```diff
@@ -4,14 +4,40 @@ extends PopochiuDialogText
...
+				current_line_number += 1
+				if rich_text_label.text[current_character-1] == " ":
+					rich_text_label.text[current_character-1] = "\n"
+				elif rich_text_label.text[current_character-1] != "\n":
+					rich_text_label.text = rich_text_label.text.left(current_character) +\
+					"\n" + rich_text_label.text.right(-current_character)
+
+				if current_line_number == number_of_lines_of_text - 1:
+					break
+	return rich_text_label.text
```
Comment (mapedorr) [inline addons/popochiu/engine/objects/gui/components/dialog_text/dialog_caption/dialog_caption.gd:41]: I like what this code does, but it shouldn't be duplicated in the 3 **DialogText** components. You should move this to the `dialog_text.gd` script so any **DialogText** component can make use of it.

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_overhead/dialog_overhead.gd [hunk 5]
```diff
@@ -44,4 +43,28 @@ func _get_icon_to_position() -> float:
...
+					rich_text_label.text[current_character-1] = "\n"
+				elif rich_text_label.text[current_character-1] != "\n":
+					rich_text_label.text = rich_text_label.text.left(current_character) + "\n" +\
+					rich_text_label.text.right(-current_character)
+
+				if current_line_number == number_of_lines_of_text - 1:
+					break
+	return rich_text_label.text
+
+
```
Comment (mapedorr) [inline addons/popochiu/engine/objects/gui/components/dialog_text/dialog_overhead/dialog_overhead.gd:69]: Remove this after moving this function to the `dialog_text.gd` script.

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_overhead/dialog_overhead.gd [hunk 6]
```diff
@@ -27,13 +26,13 @@ func _set_default_label_size(lbl: Label) -> void:
 
 
 func _append_text(msg: String, props: Dictionary) -> void:
+	msg = _correct_line_breaks(msg)
+
```
Comment (mapedorr) [inline addons/popochiu/engine/objects/gui/components/dialog_text/dialog_overhead/dialog_overhead.gd:30]: Assign the `[color][/color]` to the `msg` in the same way you did for the `dialog_caption.gd:17` and the `dialog_portrait.gd:47`.

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_portrait/dialog_portrait.gd [hunk 7]
```diff
@@ -42,4 +42,32 @@ func _set_default_size() -> void:
...
+				if rich_text_label.text[current_character-1] == " ":
+					rich_text_label.text[current_character-1] = "\n"
+				elif rich_text_label.text[current_character-1] != "\n":
+					rich_text_label.text = rich_text_label.text.left(current_character) + "\n" +\
+					rich_text_label.text.right(-current_character)
+
+				if current_line_number == number_of_lines_of_text - 1:
+					break
+	return rich_text_label.text
+
```
Comment (mapedorr) [inline addons/popochiu/engine/objects/gui/components/dialog_text/dialog_portrait/dialog_portrait.gd:72]: Remove this after moving this function to the `dialog_text.gd` script.

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_text.gd [hunk 8]
```diff
@@ -73,14 +73,18 @@ func play_text(props: Dictionary) -> void:
 	
 	if PopochiuConfig.should_talk_gibberish():
 		msg = D.create_gibberish(msg)
-	
+
 	# Call the virtual method that modifies the size of the RichTextLabel in case the dialog style
 	# requires it.
 	await _modify_size(msg, props.position)
 	
+	rich_text_label.push_color(props.color)
```
Comment (mapedorr) [inline addons/popochiu/engine/objects/gui/components/dialog_text/dialog_text.gd:81]: We removed this in #357 because it was causing issues when centering text. Please remove it again since each **DialogText** component should assign the color itself using BBCode (`[color][/color]`).

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_text.gd [hunk 9]
```diff
@@ -73,14 +73,18 @@ func play_text(props: Dictionary) -> void:
...
+	
 	# Assign the text and align mode
-	msg = "[color=%s]%s[/color]" % [props.color.to_html(), msg]
-	_append_text(msg, props)
+	if PopochiuConfig.should_talk_gibberish():
+		_append_text(D.create_gibberish(msg), props)
+	else:
+		_append_text(msg, props)
```
Comment (mapedorr) [inline addons/popochiu/engine/objects/gui/components/dialog_text/dialog_text.gd:87]: This should be just this since the gibberish text replacement was already done in line 75.

```suggestion
	_append_text(msg, props)
```

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_text.gd [hunk 10]
```diff
@@ -169,46 +173,37 @@ func _modify_size(_msg: String, _target_position: Vector2) -> void:
...
 	rich_text_label.add_child(rt)
 	
 	# Create a Label to check if the text exceeds the wrap_width
 	var lbl := Label.new()
 	lbl.add_theme_font_override("normal_font", get_theme_font("normal_font"))
 	
-	_set_default_label_size(lbl)
-	
 	lbl.text = rt.get_parsed_text()
+	lbl.size = lbl.get_minimum_size()
```
Comment (mapedorr) [inline addons/popochiu/engine/objects/gui/components/dialog_text/dialog_text.gd:191]: I think this is not necessary since the `custom_minimum_size` of the component at this point will be **Vector2.ZERO**.

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_text.gd [hunk 11]
```diff
@@ -169,46 +173,37 @@ func _modify_size(_msg: String, _target_position: Vector2) -> void:
 
 
 ## Creates a RichTextLabel to calculate the resulting size of this node once the whole text is shown.
+## Uses a RichTextLabel to provide the "get_parsed_text" function, and a label within it to work out
+## the minimum size. Calculating the size from just the RichTextLabel does not work.
```
Comment (mapedorr) [inline addons/popochiu/engine/objects/gui/components/dialog_text/dialog_text.gd:177]: If you want to replace the description of the function, please remove the previous one. Also, I suggest a small change to make it clear what the function really does.

```suggestion
## Uses a [RichTextLabel] and a [Label] to work out the minimum size since calculating it from just
## the [RichTextLabel] does not work.
```

**Review by mapedorr — CHANGES_REQUESTED** [review]
I request some changes before merging this to `develop`.

**Comment by mapedorr** [comment]
I think the `_correct_line_breaks()` function you created to modify the text so the line breaks are displayed properly works fine, and it actually improves the behavior of the tween, but I suggest making that calculation before trying to calculate the size of each **DialogText** component.

## About the Overhead dialog
The idea of this component is to take into account the `dialog_pos` property of each **PopochiuCharacter** to make the bottom of the component to render above that line. In many cases, your implementation makes the text to render below that line, which can cause reading issues on long texts.

👇 **Your implementation** 👇

https://github.com/user-attachments/assets/9c07f214-2ffe-4f89-a4f7-f21769b62a98


👇 **The current (an expected) behavior** 👇

https://github.com/user-attachments/assets/bd3df226-ca60-444b-a471-ce1a74989a5a

## About the Portrait dialog
Due to how this scene is structured, the size of the **RichTextLabel** adjusts to the available space within the **PanelContainer** component where it is contained. Perhaps what should happen with this component is that the value of the `wrap_width` property defines the size that the **PanelContainer** should have. In any case, modifying the text using your `_correct_line_breaks()` function can help the make the tween look better.

## About the Caption dialog
As I mentioned in one of the comments, I think the ideal approach for this component is that the value of the `wrap_width` property defines the X size of the **RichTextLabel** component on its `_ready()` function. The _width_ of this component will remain fixed regardless of the amount of content to render. I think modifying only the height based on the content is fine (as it is in the current implementation (in _develop_)), but we could use the `_correct_line_breaks()` function to ensure the final size adjusts better based on the final text content.

# To summarize
I think the right approach to make use of the code you made for calculating the line breaks should be used before calling `_modify_size()` (line 79 in your code) in the **DialogText** class. That way, the calculation of the size of the **RichTextLabel** component will use the text with the line breaks fixes.

File: addons/popochiu/engine/objects/gui/components/dialog_text/dialog_caption/dialog_caption.gd [hunk 12]
```diff
@@ -4,14 +4,40 @@ extends PopochiuDialogText
...
+	msg = _correct_line_breaks(msg)
 	rich_text_label.text = "[center][color=%s]%s[/color][/center]" % [props.color.to_html(), msg]
 
 
+## Appends text for the dialog caption
+## Ensures that where a printing a word would see it wrap to the next line that the newline
+## is forced into the text. Without this the tween in dialog_text.gd would print part of the word
+## until it runs out of space, then erase the part word and rewrite it on the next line which looks
+## messy.
+func _correct_line_breaks(msg: String) -> String:
```
Comment (mapedorr) [inline addons/popochiu/engine/objects/gui/components/dialog_text/dialog_caption/dialog_caption.gd:25]: I think `_fix_line_breaks` is a better name for this function. "Correct" sounds weird.

**Comment by stickgrinder** [comment]
Reviewing this with @mapedorr we agree this component may needs improvement, but the proposed implementation has some downsize.

First of all, forcefully changing the actual string may break the localization (when it will be available), since the string as it appears in the script is the key for the translation.

Also there is a lot of code which is probably better handled by the nodes internals (we should be able to rely on the standard behavior of RichTextLabels for most of the solution).

We'll keep the issue open for review, but are closing this PR.
